packages/core/src/lib/register.ts

```diff
@@ -0,0 +1,99 @@
+import { PasscodeType } from '@logto/schemas';
+import { Context } from 'koa';
+import { Provider } from 'oidc-provider';
+
+import RequestError from '@/errors/RequestError';
+import { WithUserLogContext } from '@/middleware/koa-user-log';
+import { hasUser, hasUserWithEmail, insertUser } from '@/queries/user';
+import assertThat from '@/utils/assert-that';
+import { emailReg } from '@/utils/regex';
+
+import { createPasscode, sendPasscode, verifyPasscode } from './passcode';
+import { encryptUserPassword, generateUserId } from './user';
+
+const assignRegisterResult = async (ctx: Context, provider: Provider, userId: string) => {
+  const redirectTo = await provider.interactionResult(
+    ctx.req,
+    ctx.res,
+    {
+      login: { accountId: userId },
+    },
+    { mergeWithLastSubmission: false }
+  );
+  ctx.body = { redirectTo };
+};
+
+export const registerWithUsernameAndPassword = async (
+  ctx: WithUserLogContext<Context>,
+  provider: Provider,
+  username: string,
+  password: string
+) => {
+  assertThat(
+    username && password,
+    new RequestError({
+      code: 'session.insufficient_info',
+      status: 400,
+    })
+  );
+  assertThat(
+    !(await hasUser(username)),
+    new RequestError({
+      code: 'user.username_exists',
+      status: 422,
+    })
+  );
+
+  const id = await generateUserId();
+
+  const { passwordEncryptionSalt, passwordEncrypted, passwordEncryptionMethod } =
+    encryptUserPassword(id, password);
+
+  await insertUser({
+    id,
+    username,
+    passwordEncrypted,
+    passwordEncryptionMethod,
+    passwordEncryptionSalt,
+  });
+
+  await assignRegisterResult(ctx, provider, id);
+};
+
+export const sendRegisterWithEmailPasscode = async (ctx: Context, jti: string, email: string) => {
+  assertThat(emailReg.test(email), new RequestError('user.invalid_email'));
+  assertThat(
+    !(await hasUserWithEmail(email)),
+    new RequestError({
+      code: 'user.email_exists',
+      status: 422,
+    })
+  );
+
+  const passcode = await createPasscode(jti, PasscodeType.Register, { email });
+  await sendPasscode(passcode);
+  ctx.state = 204;
+};
+
+export const registerWithEmailAndPasscode = async (
```

Comment:
will it be better to have a unified `register()` function like:

```ts
type RegisterParams = {
  idType: 'username' | 'email';
  id: string;
  codeType: 'password' | 'passcode';
  code: string;
  jti: string;
  // ...
};

const register = async ({ ... }: RegisterParams) => {
```

then we can re-use logics and be compatible with more possibilities.

Reply:
created LOG-1484 to refactor both register and sign in logics.

Reply:
ok